e3sm_to_cmip/cmor_handlers/utils.py

```diff
@@ -404,3 +375,221 @@ def _get_handler_module(module_name: str, module_path: str):
     module = SourceFileLoader(module_name, module_path).load_module()
 
     return module
+
+
+def _derive_handler(
+    var_handlers: List[Dict[str, Any]],
+    freq: Frequency,
+    realm: Union[Realm, MPASRealm],
+    cmip_tables_path: str,
+    e3sm_vars: List[str],
+) -> Optional[Dict[str, Any]]:
+    """Attempts to derive a handler for a CMIP variable.
+
+    The function first filters the handlers to those compatible with the
+    requested frequency. It then tries to find a handler whose required raw
+    E3SM variables are present in the input E3SM variable list. If no compatible
+    handler is found, and no handlers match the requested frequency, it attempts
+    to update the handler's table to match the requested frequency and tries again.
+
+    Parameters
+    ----------
+    var_handlers : List[Dict[str, Any]]
+        List of variable handler dictionaries.
+    freq : Frequency
+        The requested output frequency.
+    realm : Union[Realm, MPASRealm]
+        The realm.
+    cmip_tables_path : str
+        Path to the CMIP6 tables.
+    e3sm_vars : List[str]
+        List of available E3SM variables.
+
+    Returns
+    -------
+    Optional[Dict[str, Any]]
+        The derived handler dictionary if found, otherwise None.
+    """
+    # Step 1: Filter handlers by frequency and attempt to derive a handler.
+    filtered_handlers = _select_handlers_for_freq(var_handlers, freq)
+    derived_handler = _find_handler_by_e3sm_vars(e3sm_vars, filtered_handlers)
+
+    if derived_handler is not None:
+        return derived_handler
+
+    # Step 2: Fallback for non-monthly frequencies
+    # If no suitable handler can be derived, try updating the table for each handler
+    # and try deriving a handler again. This maintains backwards compatibility.
+    if freq != "mon":
+        adjusted_handlers = _adjust_handlers_cmip_table_for_freq(
+            var_handlers, freq, realm, cmip_tables_path
+        )
+
+        if adjusted_handlers is not None:
+            derived_handler = _find_handler_by_e3sm_vars(e3sm_vars, adjusted_handlers)
+
+            if derived_handler is not None:
+                return derived_handler
```

Comment:
Note, this is the legacy logic that existed before, which did not work properly when multiple handlers are defined for a variable. For example, with `rlut` and `huss` with different frequencies for `day` and `mon` (related to #216)